src/zope/server/trigger.py

```diff
@@ -123,106 +120,126 @@ def handle_read(self):
+            # Starting in Python 2.6, the descriptor passed to
+            # file_dispatcher gets duped and assigned to
+            # self.socket.fd. This breals the instantiation semantics and
+            # is a bug imo.  I dount it will get fixed, but maybe
+            # it will. Who knows. For that reason, we test for the
+            # fd changing rather than just checking the Python version.
+            os.close(r)
+
+    def _close(self):
+        for fd in self._fds:
+            try:
                 os.close(fd)
-            self._fds = []
+            except OSError:
+                pass
+        self._fds = []
 
-        def _physical_pull(self):
-            os.write(self.trigger, b'x')
+    def _physical_pull(self):
+        os.write(self.trigger, b'x')
 
-else:
+class BindError(Exception):
+    pass
+
+class sockettrigger(_triggerbase, asyncore.dispatcher):
     # Windows version; uses just sockets, because a pipe isn't select'able
     # on Windows.
+    kind = "loopback"
+
+    ADDR_IN_USE_CODES = (getattr(errno, 'EADDRINUSE', -1),
+                         getattr(errno, 'WSAEADDRINUSE', -1))
+
+    def __init__(self):
+        _triggerbase.__init__(self)
+
+        # Get a pair of connected sockets.  The trigger is the 'w'
+        # end of the pair, which is connected to 'r'.  'r' is put
+        # in the asyncore socket map.  "pulling the trigger" then
+        # means writing something on w, which will wake up r.
+
+        w = socket.socket()
+        # Disable buffering -- pulling the trigger sends 1 byte,
+        # and we want that sent immediately, to wake up asyncore's
+        # select() ASAP.
+        w.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+
+        count = 0
+        while 1:
+            count += 1
+            # Bind to a local port; for efficiency, let the OS pick
+            # a free port for us.
+            # Unfortunately, stress tests showed that we may not
+            # be able to connect to that port ("Address already in
+            # use") despite that the OS picked it.  This appears
+            # to be a race bug in the Windows socket implementation.
+            # So we loop until a connect() succeeds (almost always
+            # on the first try).  See the long thread at
+            # http://mail.zope.org/pipermail/zope/2005-July/160433.html
+            # for hideous details.
+            a = socket.socket()
+            a.bind(("127.0.0.1", 0))
+            connect_address = a.getsockname()  # assigned (host, port) pair
+            a.listen(1)
+            try:
+                self._connect_client(w, connect_address)
+                break    # success
+            except socket.error as detail:
+                if detail.args[0] not in self.ADDR_IN_USE_CODES:
+                    # "Address already in use" is the only error
+                    # I've seen on two WinXP Pro SP2 boxes, under
+                    # Pythons 2.3.5 and 2.4.1.
```

Comment:
Idea: git blame to discover who is the "I" in this comment and add an attribution.

(This assumes the original attribution hasn't already been lost.)

Reply:
It appears to be Tim Peters https://github.com/zopefoundation/ZEO/commit/c4f736a78ca6713fc3dec21f8aa1fa6f144dd82f